git-branchless/src/revset/eval.rs

```diff
@@ -204,14 +206,33 @@ pub(super) fn eval_name(ctx: &mut Context, name: &str) -> EvalResult {
 }
 
 pub(super) fn eval_fn(ctx: &mut Context, name: &str, args: &[Expr]) -> EvalResult {
-    let function = FUNCTIONS
-        .get(name)
-        .ok_or_else(|| EvalError::UnboundFunction {
-            name: name.to_owned(),
-            available_names: FUNCTIONS.keys().sorted().copied().collect(),
-        })?;
-
-    function(ctx, name, args)
+    if let Some(function) = FUNCTIONS.get(name) {
+        return function(ctx, name, args);
+    }
+
+    let alias_key = format!("branchless.revsets.alias.{}", name);
+    let alias_template = ctx
+        .repo
+        .get_readonly_config()
+        .map_err(EvalError::OtherError)?
+        .get::<String, String>(alias_key)
+        .map_err(EvalError::OtherError)?;
+    if let Some(mut alias_template) = alias_template {
+        for (i, arg) in args.iter().enumerate() {
+            alias_template =
+                alias_template.replace(format!("${}", i + 1).as_str(), format!("{}", arg).as_str());
+        }
+        let expr = parse(&alias_template)
+            .wrap_err("Parsing alias expression")
+            .map_err(EvalError::OtherError)?;
+        let commits = eval_inner(ctx, &expr);
+        return commits;
+    }
```

Comment:
It's worth calling attention to the ... hacky? ... crude? ... expeditious? way that I implemented this: When an unrecognized fn is encountered, we check to see if a corresponding alias is defined. If so, we serialize each element of `args` (from `Expr` back into `Strings`) and replace each placeholder in the alias "template" w/ the corresponding serialized arg. Then we re-parse the alias (w/ the placeholders replaced) and send the result back through `eval_inner`. 

I suspect that this will be fine for most cases, but if someone creates some complex aliases that also include other complex aliases, which include complex aliases, etc, etc, then this will result in a lot of serializing and re-parsing.

A more robust option might be to just parse the alias as is (w/ placeholders *not* replaced) but then pass some sort of "unbound arg resolver" callback through the eval system, and that callback could be used to resolve things like `$n` placeholders into parsed `Expr`s. I suspect that this might introduce some significant complexity into the eval system, though.

As I said, I'm assuming that the current approach will be OK for many or most use cases, but if you dislike it, I could look into an alternative approach. Or, we could go w/ this for now as an experiment and revisit it if we encounter performance issues.

Reply:
For the sake of discussion, I've pushed a fixup commit w/ an alternate approach. Instead of serializing and reparsing expressions, this latest version works by:
1. Parsing the alias template as is, w/o replacing placeholders.
2. Crawls through the parsed alias template `Expr` and replaces any names in the AST that "match" the input args (ie `$1` from the template) w/ a copy of the corresponding `Expr` from the input (ie `myAlias(foo)` would end up replacing `$1` with `Expr:Name(foo)`)
3. Evaluating the new `Expr`

I'm not familiar w/ `Cow`, so there's a lot of `.clone()` happening in the new `replace_names()` fn, and I'm not sure if that's really how it's intended to be used. 100% open to pointers, as usual. If this over complicates things, I'm happy to revert it ... or expand/alter it if you'd prefer.

Thanks for considering!

Reply:
I don't think there's any need to use `Cow` here. Most of the places I use it are for practice and to document that in principle the type might borrow e.g. from the source text, to future-proof it a little.

Reply:
The approach re. syntactic macros looks good to me 👍.